Adding "requiredExtensions" to core spec (#220)
deislabs/cnab-spec: 101-bundle-json.md

  - `EXTENSION NAME`: a unique name for an extension. Names SHOULD follow the dotted name format described earlier in this section.
  - The value of the extension must be valid JSON, but is otherwise undefined.

### Required Extensions

> Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that bundle. In this case, a bundle author MUST use the `requiredExtensions` array to define those extensions that are necessary for installing the bundle. The `requiredExtensions` array should contain the `EXTENSION NAME` defined in the `custom` object for any extensions that are required for bundle operations.

Reviewer: I'm with pedantry: "...are necessary for performing all actions with the bundle...." or some such.

> Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that bundle. In this case, a bundle author MUST use the `requiredExtensions` array to define those extensions that are necessary for installing the bundle. The `requiredExtensions` array should contain the `EXTENSION NAME` defined in the `custom` object for any extensions that are required for bundle operations.

Reviewer 2: ```suggestion
Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that extension. In this case, a bundle author MUST use the `requiredExtensions` array to define those extensions that are necessary for installing the bundle. The `requiredExtensions` array should contain the `EXTENSION NAME` defined in the `custom` object for any extensions that are required for bundle operations.
```

> Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that bundle. In this case, a bundle author MUST use the `requiredExtensions` array to define those extensions that are necessary for installing the bundle. The `requiredExtensions` object should contain the `EXTENSION NAME` defined in the `custom` object for any extensions that are required for bundle operations.

Reviewer 3: ```suggestion
Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that bundle. In this case, a bundle author MUST use the `requiredExtensions` array to define those extensions that are necessary for installing the bundle. The `requiredExtensions` array should contain the `EXTENSION NAME` defined in the `custom` object for any extensions that are required for bundle operations.
```

> Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that extension. In this case, a bundle author MUST use the `requiredExtensions` array to define those extensions that are necessary for installing the bundle. The `requiredExtensions` array should contain the `EXTENSION NAME` defined in the `custom` object for any extensions that are required for bundle operations.

Reviewer 3: Please can be get rid of the confusing "In this case" and remove the bias towards installation thus:
```suggestion
Some extensions MAY be required in order for a runtime to any action on the bundle. A bundle author MUST use the `requiredExtensions` array to define those extensions that are required. The `requiredExtensions` array SHOULD contain the `EXTENSION NAME` defined in the `custom` object for each extension that is required.
```
```suggestion
Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that extension. In this case, a bundle author MUST use the `requiredExtensions` array to define those extensions that are necessary for installing the bundle. The `requiredExtensions` array should contain the `EXTENSION NAME` defined in the `custom` object for any extensions that are required for bundle operations.
```

> A runtime SHOULD check that it supports any required custom actions before performing any operation. However, bundles SHOULD be installable by runtimes that do not support extensions that are NOT identified in the `requiredExtensions` field.  

Reviewer: Have we decided on SHOULD? For extensions, SHOULD implies that a tool could not bother to check and end up running malbehaved invocationImages -- in the sense that they were intended to be run along with a specific extension that the runtime does not support. This, I dislike, in that there's no statement here that would create a requirement to indicate to the user that it cannot support the bundle. 

If I am reading this correctly, I would urge MUST on our tools here. Even non-supporting tools would then give the user an opportunity to do something different. With SHOULD, we seem to be saying `meh`. Do I misunderstand?

Reviewer: I'd very much like a statement SHOULD about the failure case for tooling. If a run time does not support any specified extension -- or extensions in general, should those different cases be called out? -- I'd very much like to emphasize that the proper way to handle this is to return as much information about the extensions not supported or about the extension feature itself so that users can have something to deal with as a result of the failure. This may be obvious, but asserting a SHOULD is a great emphasis for tool builders here.

Reviewer 3: I agree with @squillace here, but I wonder why the text switches from talking about required extensions to required custom actions. How about the following based on @jeremyrickard's suggested text?
```suggestion
A runtime MUST check that it supports any required extensions before performing any operation. If the runtime does not support the required extension(s), it MAY proceed with the operation or fail, however it MUST notify the user that it does not support the required extension(s). Bundles SHOULD be installable by runtimes that do not support extensions that are NOT identified in the `requiredExtensions` field. 
```

  - `EXTENSION NAME`: a unique name for an extension. Names SHOULD follow the dotted name format described earlier in this section.
  - The value of the extension must be valid JSON, but is otherwise undefined.

### Required Extensions

> Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that bundle. In this case, a bundle author MUST use the `requiredExtensions` object to define those extensions that are necessary for installing the bundle. The `requiredExtensions` object should contain a collection of key/value pairs. The key MUST be the `EXTENSION NAME` and the value MUST be a well known JSON Schema reference that can be used to validate the JSON data contained in the `custom` object.

Author: I'm good with all of this, except:

> I'm not sure what "well known" means, but I think we should require a URL for the schema reference as the runtime needs to be able to locate (and "download") the schema not just know it's identifier.

I don't think we can assume that a tool can ever "download" the schema at runtime. In the case of running the tool in a disconnected/offline environment, that wouldn't be possible. Most extensions are very likely to require some runtime capabilities (dependencies for example) that need to be included in the tool (it's possible some tools could dynamically load extensions but let's assume they don't for simplicity) and known ahead of time. That's what was implied by the "well known" aspect of it. I think we can absolutely refer to the _URL_ of a schema for an extension but I think we should not assume that it's going to be downloaded. Thoughts?

> Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that bundle. In this case, a bundle author MUST use the `requiredExtensions` object to define those extensions that are necessary for installing the bundle. The `requiredExtensions` object should contain a collection of key/value pairs. The key MUST be the `EXTENSION NAME` and the value MUST be a well known JSON Schema reference that can be used to validate the JSON data contained in the `custom` object.

Reviewer 3: Yes, that seems reasonable. So "well known" kind of means "globally unique" in that it somehow won't clash with the schema reference of another extension. Right?

> Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that bundle. In this case, a bundle author MUST use the `requiredExtensions` object to define those extensions that are necessary for installing the bundle. The `requiredExtensions` object should contain a collection of key/value pairs. The key MUST be the `EXTENSION NAME` and the value MUST be a well known JSON Schema reference that can be used to validate the JSON data contained in the `custom` object.

Author: I think that's a true statement. And that "cnab.io/spec/extensions/dependencies" refers to dependencies and that a bundle author can say "i need that extension" and tools that support dependencies will work with that reference. Maybe we don't really need the reference, maybe we need agreed upon extension names and it's just assumed that tooling will know that extension name and support it or not. I think that's really what I was trying to get at with that.

> A runtime MUST validate that it supports any defined custom actions before performing any operation. However, bundles SHOULD be installable by runtimes that do not understand the extensions if they are NOT identified in the `requiredExtensions` field.  

Reviewer 3: 1. Weaken "MUST" to "SHOULD" for consistency with the above comment on allowing a user to "force" an operation.

2. Perhaps we should avoid the word "validate" here as it might imply just schema validation.

3. We should talk about "required" actions rather than "defined" actions.

4. Let's use "support" rather than "understand" for consistency.
```suggestion
A runtime SHOULD check that it supports any required custom actions before performing any operation. However, bundles SHOULD be installable by runtimes that do not support extensions that are NOT identified in the `requiredExtensions` field.  
```

Reviewer: My bad, @glyn, related to my comment https://github.com/deislabs/cnab-spec/pull/220#discussion_r302651720. I intended in my comment that MUST should be applied to the validation step, but NOT to the `force` activity step. For example, I would suggest that a runtime MUST inform the user that it doesn't know how to perform the custom extension X, but if the user passes a `--force` argument that it will perform the action in any case because "the user knows best". 

If that makes sense to us all, how should we state this? Maybe something like, "A runtime MUST check that it supports any required custom action... and it MUST require an explicit override statement if the user insists on attempting to perform an action on the bundle." I'm open to wording here, as my goal is to break apart the MUST CHECK AND STOP portion from the SHOULD ENABLE OVERRIDE portion.

> See [101.03-bundle.json](examples/101.03-bundle.json) for a complete example.

Reviewer 3: Elsewhere, we inline the example and follow it with "Source: [link]".
